### DISCUSSION TITLE
Prototype based inheritance and ClrTypes

### DISCUSSION THREAD

**Discussion by LUUUYYUF** [sebastienros/jint#1869] [Q&A] [answered]
Hello

I have a lot of old code that uses prototype base inheritance and I need to rewrite some of it to C#.

.NET:
```
class A
{
	public void testA() => Console.WriteLine("testA");
}
var engine = new Engine(opt =>
{
	opt.Strict = false;
	opt.AllowClr();
});

engine.SetValue("A", TypeReference.CreateTypeReference<A>(engine));
engine.Execute(`JavaScript Code`);
```

Javascript code using `class` works fine, I can call `testA` method.

```
class C extends A {
  constructor() {
    super();
    this.initialize(...arguments);
  }
  initialize() {
  __Log('C initialize');
  }
}
var obj = new C();
obj.testA();
```

But prototype version doesn't work.
I get error `Jint.Runtime.JavaScriptException: Property 'testA' of object is not a function`

```
function B() {
    this.initialize(...arguments);
}

B.prototype = Object.create(A.prototype);
B.prototype.constructor = B;
B.prototype.initialize = function() {
    __Log('B initialize');
};
var obj = new B();
obj.testA();
```

I don't know what to do about it, please give me some guidance.
I can't rewrite entire code to use `class`.

Thanks!

**Comment by lahma** [comment]
Was this working with some Jint version?

**Reply by LUUUYYUF** [reply]
Nuget versions: from 3.1.1 to 3.0.0 - same result.
2.11.58 - parse of java script code fails ` Jint.Parser.ParserException: Line 3: Unexpected token .`

**Reply by lahma** [reply]
So this has never worked and you haven't used Jint for this before?

**Reply by LUUUYYUF** [reply]
Never worked and I'm first time using Jint. 
I checked that it is correct javascript code.

**Reply by lahma** [reply, marked as answer]
OK then, you are probably the first one to try this way then and Jint just doesn't support to prototype approach at the moment.

**Reply by lahma** [reply]
The class based one was implemented here https://github.com/sebastienros/jint/pull/1168 , not sure if it will give some pointers how the prototype version support could be implemented.

**Reply by LUUUYYUF** [reply]
Unfortunately, I don't have the resources to devote to extra analysis and development.
I will solve that problem by creating proxy classes in javascript which will be using Clr Types directly.
No javascript classes will inherit from .net classes.
Thank You for your help.

**Reply by lahma** [reply]
Due to how Jint handles type references and expects things to be static as it doesn't currently know the target, I guess your example would work if you would make `testA()` a static method. But that might not work for your use case.
